### PR TITLE
[SPARK-38102][CORE] Support custom commitProtocolClass in saveAsNewAPIHadoopDataset

### PR DESCRIPTION
**Pull request by ocworld**
### What changes were proposed in this pull request?
This is a follow-up PR to [[SPARK-38102](https://issues.apache.org/jira/browse/SPARK-38102)]

It is needed to support custom commitProtocolClass class when using saveAsNewAPIHadoopDataset by an option. For example,

spark.hadoop.mapreduce.sources.commitProtocolClass org.apache.spark.internal.io.cloud.PathOutputCommitProtocol

### Why are the changes needed?
There is no way to apply spark-hadoop-cloud's commitProtocolClass when using saveAsNewAPIHadoopDataset that are to avoid object storage's problem.

https://spark.apache.org/docs/latest/cloud-integration.html

### Does this PR introduce _any_ user-facing change?
Yes. The option is added that is spark.hadoop.mapreduce.sources.commitProtocolClass

### How was this patch tested?
1. Adding unittest
2. Tested in my cluster

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/FileSuite.scala [hunk 1]
```diff
@@ -485,6 +486,23 @@ class FileSuite extends SparkFunSuite with LocalSparkContext {
...
+      Seq(("key1", "a"), ("key2", "a"), ("key3", "b"), ("key4", "c")), 1)
+    val job = Job.getInstance(sc.hadoopConfiguration)
+    job.setOutputKeyClass(classOf[String])
+    job.setOutputValueClass(classOf[String])
+    job.setOutputFormatClass(classOf[NewTextOutputFormat[String, String]])
+    val jobConfig = job.getConfiguration
+    jobConfig.set("mapreduce.output.fileoutputformat.outputdir",
+      tempDir.getPath + "/outputDataset_new")
+    jobConfig.set("mapreduce.sources.commitProtocolClass",
+      classOf[HadoopMapReduceCommitProtocol].getName)
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/FileSuite.scala:501]: Does this test case fail without your patch? If not, this is invalid.

File: core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala [hunk 2]
```diff
@@ -339,7 +339,8 @@ class HadoopMapReduceWriteConfigUtil[K, V: ClassTag](conf: SerializableConfigura
 
   override def createCommitter(jobId: Int): HadoopMapReduceCommitProtocol = {
     FileCommitProtocol.instantiate(
-      className = classOf[HadoopMapReduceCommitProtocol].getName,
+      className = getConf.get("mapreduce.sources.commitProtocolClass",
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala:342]: One workaround might be:

```scala
    val config = new HadoopMapReduceWriteConfigUtil[K, V](new SerializableConfiguration(conf)) {
      override def createCommitter(jobId: Int): HadoopMapReduceCommitProtocol = {
        FileCommitProtocol.instantiate(
          className = "CustomClassName",
          jobId = jobId.toString,
          outputPath = new SerializableConfiguration(
            conf).value.get("mapreduce.output.fileoutputformat.outputdir")
        ).asInstanceOf[HadoopMapReduceCommitProtocol]
      }
    }
    SparkHadoopWriter.write(
      rdd = self,
      config = config)
```

Although `HadoopMapReduceWriteConfigUtil` would have to be inherited within `org.apache.spark` package in your project though (if you're using Scala). For Java, it won't validate the package name.

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala:342]: The problem here is that we're super conservative in RDD and core API now, and won't likely add something new except bug fixes.

Comment (ocworld) [reply core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala:342]: > The problem here is that we're super conservative in RDD and core API now, and won't likely add something new except bug fixes.

Okay. I understand about it. Thank you for your comment.  I will try to your suggested way. Fortunately, I am using scala.

**Pull request closed without merging**